File: javascript/src/ExpressionFactory.ts
import CucumberExpression from './CucumberExpression.js'
import type ParameterTypeRegistry from './ParameterTypeRegistry.js'
import RegularExpression from './RegularExpression.js'
import type { Expression } from './types.js'

export default class ExpressionFactory {
  public constructor(private readonly parameterTypeRegistry: ParameterTypeRegistry) {}

  public createExpression(expression: string | RegExp): Expression {
    return typeof expression === 'string'
      ? new CucumberExpression(expression, this.parameterTypeRegistry)
      : new RegularExpression(expression, this.parameterTypeRegistry)
  }
}

File: dart/lib/src/expression_factory.dart
import 'package:cucumber_expressions/src/cucumber_expression.dart';
import 'package:cucumber_expressions/src/expression.dart';
import 'package:cucumber_expressions/src/parameter_type_registry.dart';
import 'package:cucumber_expressions/src/regular_expression.dart';

/// Creates [Expression]s from either a [String] Cucumber Expression or a
/// [RegExp] regular expression.
class ExpressionFactory {
  /// Creates a factory backed by [_parameterTypeRegistry].
  ExpressionFactory(this._parameterTypeRegistry);

  final ParameterTypeRegistry _parameterTypeRegistry;

  /// Creates an [Expression] from [expression].
  ///
  /// [expression] must be a [String] or a [RegExp]; otherwise an
  /// [ArgumentError] is thrown.
  Expression createExpression(Object expression) {
    if (expression is RegExp) {
      return RegularExpression(expression, _parameterTypeRegistry);
    }
    if (expression is String) {
      return CucumberExpression(expression, _parameterTypeRegistry);
    }
    throw ArgumentError(
      'Expression must be a String or a RegExp, was ${expression.runtimeType}',
    );
  }
}
